### PR TITLE
Add Model.topology to provide labels for observability

### PR DESCRIPTION
**Pull request by PietroPasotti**
Lots of charm code relies on https://github.com/canonical/prometheus-k8s-operator/blob/main/lib/charms/observability_libs/v0/juju_topology.py to extract a 'juju topology' object, providing a way to uniquely identify (in a human-readable manner) a unit across your juju deployments.

This PR adds to Model a `topology` property exposing the same data from native ops code.

## Checklist

 - [y] Have any types changed? If so, have the type annotations been updated?
 - [y] If this code exposes model data, does it do so thoughtfully, in a way that aids understanding?
 - [n/a] Do error messages, if any, present charm authors or operators with enough information to solve the problem?

## QA steps

Added a test verifying the API.

## Documentation changes
None necessary.

## Changelog

*Please include a bulleted list of items here, suitable for inclusion in a release changelog.*

- Added Model.topology to expose juju topology information to uniquely identify the current unit.

### CODE REVIEW

File: ops/model.py [hunk 1]
```diff
@@ -259,6 +269,25 @@ def get_binding(self, binding_key: Union[str, 'Relation']) -> Optional['Binding'
...
+    @property
+    def topology(self) -> '_Topology':
+        """A dict uniquely identifying this unit across all your Juju-managed deployments.
+
+        This is like a fingerprint that combines the following four elements:
+        - ``name``: Model name
+        - ``model_uuid``: Model UUID
+        - ``application``: Application name
+        - ``unit``: Unit name
+        - ``charm_name``: The name of the charm
```
Comment (simskij) [inline ops/model.py:281]:
```suggestion
        """Topological metadata allowing to identify a node across a set of Juju-managed deployments.
        
        The Juju topology allows for identifying unique nodes across multiple levels of granularity using different combinations of the topological components, in particular:
        - models using `model` and `model_uuid`
        - applications using `model`, `model_uuid`, and `application`
        - units using `model`, `model_uuid`, `application`, and `unit`
        - charms using `charm_name` 
```

**Review by simskij — APPROVED** [review]

**Review by jnsgruk — APPROVED** [review]
I think this makes sense. It's relatively little code, and given the number of charms that will integrate with the observability stack, it seems sensible to add this provision into ops itself

**Comment by benhoyt** [comment]
I realize this is a small amount of code and would be easy to add, but I'm not convinced the framework is the right place for this. I know @simskij and co would like to have this as soon as possible, but I'd like to use this as something of a test case for how we decide what should be in the framework and what should live elsewhere. I'd like to take some time to understand the context, how charm libs work, and what the alternatives are.

I understand this is going to be used for observability, so we can include charm and model/app/unit details when logging and recording metrics. That definitely seems to me like an observability concern, not a Juju model / framework concern, so it should go in a charm or charm library. I realize we want basically all our charms to have common attributes here, but that doesn't mean other charms and other organizations/groups will want the exact same set of attributes, or will do observability the same way. I think we want Ops to be the *framework* used by a wide swath of charms including community charms with different needs.

I need to come more up to speed on charm libs, but it seems to me we could fairly easily put this in a charm lib. Either copying and pasting these handful of lines of code to the 5 or 6 observability charm libs that need it (shouldn't be hard to manage with that small a number), or place this in a "topology" charm lib and include and import that manually in charms.

**Oh wait,** I just saw it already does exist as a charm lib, [juju_topology](https://github.com/canonical/observability-libs/blob/main/lib/charms/observability_libs/v0/juju_topology.py). What machinery does one use today to include a charm lib in a charm? Is it a one-liner addition to a metadata file or an import somewhere, or is there some pain that I'm missing to include this lib in charms that need it?

-----

One specific suggestion on this code (related to @simskij's suggestion): this is currently documented as a "fingerprint" and "uniquely-identifying", but it's not really a fingerprint or unique, as it contains redundant information (model + model UUID, unit name includes application name). So I think wherever we end up adding this we should be a bit clearer about the documentation, and explicit about the intended use cases.

**Comment by jnsgruk** [comment]
> **Oh wait,** I just saw it already does exist as a charm lib, [juju_topology](https://github.com/canonical/observability-libs/blob/main/lib/charms/observability_libs/v0/juju_topology.py?rgh-link-date=2022-11-05T14%3A13%3A54Z). What machinery does one use today to include a charm lib in a charm? Is it a one-liner addition to a metadata file or an import somewhere, or is there some pain that I'm missing to include this lib in charms that need it?

Indeed. You would pull the library in using `charmcraft fetch-lib <lib>`, which works fine, and can even make updating the lib fairly easy.

My perspective on why this particular piece should probably be included, is that all of the Observability charms depend on this, but also all of the charms that integrate with our Observability charms will also depend on it.

It was originally inlined in each of the o11y libs (like the prometheus lib, loki lib, etc.) but that made maintenance not a lot of fun - and because charm libs don't have dependency resolution, this has to be pulled in separately from the other libs that essentially depend on it.

**Comment by jnsgruk** [comment]
> 1. Make the o11y libs import and depend on the juju_topology lib (but manually, as charm libs don't have automatic dependencies). In other words, a charm that use an o11y lib has to fetch the o11y lib, but also `charmcraft fetch-lib charms.observability_libs.v0.juju_topology`. This is slightly annoying, but not exactly onerous, and allows easy updating if needed.

This is essentially where we're at now -- not a huge burden I agree, more a minor annoyance the first time a dev imports one of the o11y libs.

> 3\. Improve charm lib dependency management. Perhaps this has been vetoed and discussed to death already, I'm not sure? But I'm thinking something very lightweight like adding `LIBDEPS = ['charms.observability_libs.v0.juju_topology']`, and then `charmcraft fetch-lib` pulls it in.

We've been around this buoy a few times, and I know @niemeyer is not keen on adding a dependency system -- tracking cyclic dependencies and clashes gets complicated, and I think contravenes the original intent of libraries.

Not sure exactly how to progress -- do @jameinel or @niemeyer have opinions here?

**Comment by jameinel** [comment]
So I don't like adding this *just* for the purposes of one set of charms. But if we feel that we have an abstractly useful way of modeling something for charmers, that is the sort of thing to bring into OF. I'd think of it a bit like they say of Python standard lib, where "stable means dead". (eg, we shouldn't want to be updating this content, because lots of other things are going to start building on top of it.)

That said, this doesn't feel super useful generally, other than the property that most "good" charms should integrate with COS and COS is using this.

This is quite small, and so would feel very strange to be a full pip 'requirements.txt' style dependency. Though if it is stable enough to be in OF, and small enough to be a single function, it doesn't feel like it is a huge burden to copy it into the COS relation interface(s). I know it took a while to evolve, but at this point, if it isn't changing, it is only about 20 lines copied between them.

If w e did include it, I don't think it should be on *Model* as much as something like Unit. (eg, something like self.unit.fingerprint(), or unique_identifier())

**Comment by benhoyt** [comment]
So I've thought about this further and come up with a framework of sorts for deciding whether or not something should go in the Python Operator Framework (a "framework framework"!).

Here's the way I think we should view it: inclusion shouldn't be seen as a bar to reach like "is this good enough?" or "will a lot of charms use this?", rather, we should ask "is this a Juju modeling concern?" or "is this a Juju or charm concept that we need to map in the Python Operator Framework?"

Almost all the existing classes and functions in Operator Framework falls into that category -- they represent first-class Juju concepts: units, applications, status, storage, containers/Pebble interaction, relations and relation data, and soon secrets.

But "helper function to get several attributes for an observability fingerprint" does not fall into the category for inclusion.

Put another way, Python Operator Framework currently lives up to its name: a framework, not a charm standard library. The above is the general rule I think we should use for deciding.

-----

Another angle on this specific change: I've looked at the observability libs that use the existing [juju_topology](https://github.com/canonical/observability-libs/blob/main/lib/charms/observability_libs/v0/juju_topology.py) charm library, and I don't think this PR's `topology` property would be enough anyway. The `juju_topology` lib is a lot more than this, with `from_charm` and `from_dict` and the rich `JujuTopology` object that has `as_dict` (complete with ability to remap key names and exclude keys), `identifier`, and the somewhat-tricky `label_matcher_dict`.

The [loki-k8s-operator](https://github.com/canonical/loki-k8s-operator) and [prometheus-k8s-operator](https://github.com/canonical/prometheus-k8s-operator) charms use `juju_topology`, but then add even more on top of that, such as [formatting key names](https://github.com/canonical/loki-k8s-operator/blob/0e99031e1fd4785dae4bcb4027a3295ed288694f/lib/charms/loki_k8s/v0/loki_push_api.py#L2102-L2106) and [filtering keys](https://github.com/canonical/prometheus-k8s-operator/blob/05e7fb6517dcdae052db16be12510a6bbe986854/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L2154-L2159).

So even if we did include this change, those charms libraries would still need a bunch of key-formatting code and something like `label_matcher_dict`. I think `JujuTopology` is trying to do too much, and my recommendation is that the observability libs should just get the attributes they need in the correct format as they need them. This would avoid the filtering/formatting gymnastics, and also avoid the need for the `juju_toplology` dependency. For example:

```python
# Which is easier to understand? This, which we have currently:
self.topology = JujuTopology.from_charm(charm)
...
common_labels = {
    "juju_{}".format(k): v
    for k, v in self.topology.as_dict(remapped_keys={"charm_name": "charm"}).items()
}

# Or this, which is a bit repetitive, but very direct and clear:
common_labels = {
    "juju_model": self.model.name,
    "juju_model_uuid": self.model.uuid,
    "juju_application": self.model.app.name,
    "juju_unit": self.model.unit.name,
    "juju_charm": self.meta.name,
}
```

-----

With those two things in mind I think I'll be rejecting this PR, but I'll leave it open for another week or two for any further discussion.

**Comment by benhoyt** [comment]
Closing this per my [most recent comment](https://github.com/canonical/operator/pull/849#issuecomment-1316279183) above.

**Pull request closed without merging**